File: Compiler/Parser/Pattern.py
import Compiler.Syntax.Pattern as Pattern
import Compiler.Token as Token

def parsePattern(parser):
    if parser.peek(Token.Wildcard()):
        parser.expect(Token.Wildcard())
        return Pattern.Wildcard()
    elif parser.peek("typeid"):
        p = Pattern.Named()
        p.name = parser.parseTypeName()
        if parser.peek(Token.LeftParen()):
            parser.expect(Token.LeftParen())
            while True:
                arg = parsePattern(parser)
                p.args.append(arg)
                if parser.peek(Token.Comma()):
                    parser.expect(Token.Comma())
                else:
                    break
            parser.expect(Token.RightParen())
        return p
    elif parser.peek(Token.LeftParen()):
        p = Pattern.Tuple()
        if parser.peek(Token.LeftParen()):
            parser.expect(Token.LeftParen())
            while True:
                arg = parsePattern(parser)
                p.args.append(arg)
                if parser.peek(Token.Comma()):
                    parser.expect(Token.Comma())
                else:
                    break
            parser.expect(Token.RightParen())
        return p
    elif parser.peek("mut"):
        parser.step()
        p = Pattern.Bind()
        p.name = parser.parseName()
        p.mutable = True
        return p
    elif parser.peek("varid"):
        p = Pattern.Bind()
        p.name = parser.parseName()
        return p
    else:
        parser.expect("<pattern>")
        

File: Siko/src/siko/parser/Pattern.rs
use crate::siko::syntax::Pattern::{Pattern, SimplePattern};

use super::{
    Parser::Parser,
    Token::{BracketKind, KeywordKind, MiscKind, TokenKind},
};

pub trait PatternParser {
    fn buildPattern(&mut self, p: SimplePattern) -> Pattern;
    fn parsePattern(&mut self) -> Pattern;
}

impl PatternParser for Parser {
    fn buildPattern(&mut self, p: SimplePattern) -> Pattern {
        Pattern {
            pattern: p,
            location: self.popSpan(),
        }
    }

    fn parsePattern(&mut self) -> Pattern {
        self.pushSpan();
        match self.peek() {
            TokenKind::Keyword(KeywordKind::Mut) => {
                self.expect(TokenKind::Keyword(KeywordKind::Mut));
                let name = self.parseVarIdentifier();
                self.buildPattern(SimplePattern::Bind(name, true))
            }
            TokenKind::VarIdentifier => {
                let name = self.parseVarIdentifier();
                self.buildPattern(SimplePattern::Bind(name, false))
            }
            TokenKind::TypeIdentifier => {
                let name = self.parseTypeIdentifier();
                let mut args = Vec::new();
                if self.check(TokenKind::LeftBracket(BracketKind::Paren)) {
                    self.expect(TokenKind::LeftBracket(BracketKind::Paren));
                    while !self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                        let arg = self.parsePattern();
                        args.push(arg);
                        if self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                            break;
                        }
                        self.expect(TokenKind::Misc(MiscKind::Comma));
                    }
                    self.expect(TokenKind::RightBracket(BracketKind::Paren));
                }
                self.buildPattern(SimplePattern::Named(name, args))
            }
            TokenKind::LeftBracket(BracketKind::Paren) => {
                let mut args = Vec::new();
                if self.check(TokenKind::LeftBracket(BracketKind::Paren)) {
                    self.expect(TokenKind::LeftBracket(BracketKind::Paren));
                    while !self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                        let arg = self.parsePattern();
                        args.push(arg);
                        if self.check(TokenKind::RightBracket(BracketKind::Paren)) {
                            break;
                        }
                        self.expect(TokenKind::Misc(MiscKind::Comma));
                    }
                    self.expect(TokenKind::RightBracket(BracketKind::Paren));
                }
                self.buildPattern(SimplePattern::Tuple(args))
            }
            TokenKind::Misc(MiscKind::Wildcard) => {
                self.expect(TokenKind::Misc(MiscKind::Wildcard));
                self.buildPattern(SimplePattern::Wildcard)
            }
            kind => self.reportError2("<pattern>", kind),
        }
    }
}
